File: test/suites/dxfstrict.js
'use strict';
/* ============================================================
   What AutoCAD is strict about

   ezdxf reads almost anything, which is why every DXF check
   here passed while AutoCAD refused the same file outright:

     Class separator for class AcDbDimStyleTable expected on
     line 828. Invalid or incomplete DXF input — drawing
     discarded.

   The DIMSTYLE table is the one symbol table in DXF whose
   header carries a SECOND subclass marker after the entry
   count — 100 AcDbDimStyleTable, then 71 with the number of
   entries. Every other table stops at AcDbSymbolTable. Ours
   stopped there too, so AutoCAD hit the first record where it
   expected the marker and discarded the whole drawing.

   These are structural assertions rather than round trips: a
   reader that accepts our file proves nothing about a reader
   that does not.
   ============================================================ */
const SETUP = `
  resetDoc(); ensureLayer('A-WALL');
  begin();
  addEnt({t:'wall', a:[0,0], b:[6000,0], wt:'gen100', layer:'A-WALL'});
  addEnt({t:'dim', k:'linear', p1:[0,0], p2:[6000,0], off:900, layer:'DIMENSIONS'});
  addEnt({t:'text', p:[0,1200], s:'HELLO', h:200, layer:'TEXT'});
  commit('w');
`;
/** the DXF as [code, value] pairs, which is the only way to read one */
const pairsOf = (dxf) => {
  const L = dxf.split(/\r\n|\n/);
  const out = [];
  for (let i = 0; i + 1 < L.length; i += 2) out.push([L[i].trim(), L[i + 1]]);
  return out;
};
/** the group codes of one table's header, up to its first record */
const tableHead = (dxf, name) => {
  const p = pairsOf(dxf);
  for (let i = 0; i < p.length; i++) {
    if (p[i][0] === '0' && p[i][1] === 'TABLE' && p[i + 1] && p[i + 1][1] === name) {
      const head = [];
      for (let j = i + 2; j < p.length; j++) {
        if (p[j][0] === '0') break;                 /* the first record starts here */
        head.push(p[j]);
      }
      return head;
    }
  }
  return null;
};
module.exports = ({ group, t, ok, eq, close, R }) => {

  group('the tables section');

  t('the DIMSTYLE table header carries the class marker AutoCAD demands', () => {
    const dxf = R(`${SETUP} return exportDXF();`);
    const head = tableHead(dxf, 'DIMSTYLE');
    ok(head, 'there is a DIMSTYLE table');
    const flat = head.map(x => x[0] + '=' + x[1]);
    ok(flat.indexOf('100=AcDbSymbolTable') >= 0, 'the ordinary marker: ' + flat.join(' '));
    ok(flat.indexOf('100=AcDbDimStyleTable') >= 0,
      'and the one only this table has: ' + flat.join(' '));
    const i70 = flat.findIndex(x => x.slice(0, 3) === '70=');
    const iCls = flat.indexOf('100=AcDbDimStyleTable');
    ok(i70 >= 0 && iCls > i70, 'in that order — count first, then the marker');
    ok(flat.some(x => x.slice(0, 3) === '71='), 'with its own entry count after it');
  });

  t('and no other table grew one, because no other table has one', () => {
    const dxf = R(`${SETUP} return exportDXF();`);
    const bad = [];
    for (const n of ['VPORT', 'LTYPE', 'LAYER', 'STYLE', 'VIEW', 'UCS', 'APPID', 'BLOCK_RECORD']) {
      const head = tableHead(dxf, n);
      if (!head) continue;
      if (head.some(x => x[0] === '100' && x[1] !== 'AcDbSymbolTable')) bad.push(n);
    }
    eq(bad.length, 0, 'they stop at AcDbSymbolTable: ' + bad.join(', '));
  });

  t('every table is opened and closed, and every record sits inside one', () => {
    const dxf = R(`${SETUP} return exportDXF();`);
    const p = pairsOf(dxf);
    let depth = 0, worst = 0, ends = 0, opens = 0;
    for (const [c, v] of p) {
      if (c !== '0') continue;
      if (v === 'TABLE') { depth++; opens++; }
      else if (v === 'ENDTAB') { depth--; ends++; }
      if (depth < 0) worst = -1;
    }
    eq(worst, 0, 'no ENDTAB without a TABLE');
    eq(depth, 0, 'and every TABLE is closed');
    eq(opens, ends, opens + ' tables, ' + ends + ' ends');
  });

  t('the DIMSTYLE record is keyed on 105, not 5, as that table alone is', () => {
    const dxf = R(`${SETUP} return exportDXF();`);
    const p = pairsOf(dxf);
    const i = p.findIndex((x, k) => x[0] === '0' && x[1] === 'DIMSTYLE' &&
      p.slice(0, k).some(y => y[0] === '2' && y[1] === 'DIMSTYLE'));
    ok(i > 0, 'the record is there');
    const codes = p.slice(i + 1, i + 4).map(x => x[0]);
    ok(codes.indexOf('105') >= 0, 'handled with 105: ' + codes.join(','));
    eq(codes.indexOf('5'), -1, 'and not with 5');
  });

  group('the shape of the file');

  t('every section is opened and closed', () => {
    const dxf = R(`${SETUP} return exportDXF();`);
    const p = pairsOf(dxf);
    let open = 0, close = 0;
    for (const [c, v] of p) {
      if (c !== '0') continue;
      if (v === 'SECTION') open++;
      if (v === 'ENDSEC') close++;
    }
    eq(open, close, open + ' sections opened, ' + close + ' closed');
    eq(p[p.length - 1][1], 'EOF', 'and the file ends where it says it does');
  });

  t('it is R2000, and says so once', () => {
    const dxf = R(`${SETUP} return exportDXF();`);
    const n = (dxf.match(/\$ACADVER/g) || []).length;
    eq(n, 1, 'one version stamp');
    ok(dxf.indexOf('AC1015') >= 0, 'and it is R2000');
  });
  group('nothing references what the file does not define');

  /* An entity naming a layer the file never defines loses its layer in
     whatever opens it — AutoCAD drops it onto 0. The document is allowed to
     get into that state (AUDIT exists to find it), so the exporter has to be
     the one that does not pass it on. */
  t('every layer an entity names is in the LAYER table', () => {
    const dxf = R(`${SETUP}
      /* a layer nobody ensured — exactly what AUDIT reports on */
      begin();
      addEnt({t:'line', a:[0,0], b:[1000,0], layer:'A-NEVER-MADE'});
      commit('l');
      return exportDXF();`);
    const p = pairsOf(dxf);
    const defined = new Set();
    let inLayerTable = false;
    for (let i = 0; i < p.length; i++) {
      if (p[i][0] === '0' && p[i][1] === 'TABLE' && p[i+1] && p[i+1][1] === 'LAYER') inLayerTable = true;
      else if (p[i][0] === '0' && p[i][1] === 'ENDTAB') inLayerTable = false;
      else if (inLayerTable && p[i][0] === '0' && p[i][1] === 'LAYER') {
        for (let j = i + 1; j < p.length && p[j][0] !== '0'; j++)
          if (p[j][0] === '2') { defined.add(p[j][1]); break; }
      }
    }
    const used = new Set();
    for (let i = 0; i < p.length; i++) if (p[i][0] === '8') used.add(p[i][1]);
    const missing = [...used].filter(x => !defined.has(x));
    eq(missing.length, 0, 'these are drawn on but never defined: ' + missing.join(', '));
    ok(defined.has('A-NEVER-MADE'), 'including one the document never made');
  });

  t('the layer table says how many records it holds', () => {
    const dxf = R(`${SETUP}
      begin(); addEnt({t:'line', a:[0,0], b:[1,0], layer:'A-NEVER-MADE'}); commit('l');
      return exportDXF();`);
    const head = tableHead(dxf, 'LAYER');
    const n70 = head.filter(x => x[0] === '70').map(x => +x[1])[0];
    const p = pairsOf(dxf);
    let count = 0, inside = false;
    for (let i = 0; i < p.length; i++) {
      if (p[i][0] === '0' && p[i][1] === 'TABLE' && p[i+1] && p[i+1][1] === 'LAYER') inside = true;
      else if (inside && p[i][0] === '0' && p[i][1] === 'ENDTAB') break;
      else if (inside && p[i][0] === '0' && p[i][1] === 'LAYER') count++;
    }
    eq(n70, count, 'the count matches the records: ' + n70 + ' vs ' + count);
  });
};

File: tools/dxf_strict.py
#!/usr/bin/env python3
"""Check a DXF against the structural rules a STRICT reader enforces.

check_dxf.py asks ezdxf whether the drawing survives a round trip. ezdxf is
forgiving by design, and that is the gap this file exists to close: it read our
files perfectly happily while AutoCAD refused to open them at all --

    Class separator for class AcDbDimStyleTable expected on line 828.
    Invalid or incomplete DXF input -- drawing discarded.

So this reads the raw tag stream and checks the things a strict reader checks,
with no library in between. It knows nothing about geometry; it is only asking
whether the file is put together the way the format says.

    python tools/dxf_strict.py [file.dxf]
"""
import sys, collections

path = sys.argv[1] if len(sys.argv) > 1 else 'test/out/fixture.dxf'
raw = open(path, 'r', encoding='utf-8', errors='replace').read()
lines = raw.split('\r\n') if '\r\n' in raw else raw.split('\n')
while lines and lines[-1] == '':
    lines.pop()

problems = []
def bad(msg, line=None):
    problems.append(("line %d: " % line if line else "") + msg)

# ---------------------------------------------------------------- tag stream
if len(lines) % 2:
    bad("odd number of lines: a DXF is pairs of code and value, so one is missing")
pairs = []          # (code, value, line-number-of-the-code)
for i in range(0, len(lines) - 1, 2):
    c = lines[i].strip()
    try:
        code = int(c)
    except ValueError:
        bad("group code %r is not a number" % c, i + 1)
        continue
    pairs.append((code, lines[i + 1], i + 1))

# ---------------------------------------------------------------- sections
depth, sections, order = 0, [], []
for code, val, ln in pairs:
    if code != 0:
        continue
    if val == 'SECTION':
        depth += 1
        if depth > 1:
            bad("SECTION opened inside another one", ln)
    elif val == 'ENDSEC':
        depth -= 1
        if depth < 0:
            bad("ENDSEC without a SECTION", ln)
            depth = 0
if depth:
    bad("%d section(s) never closed" % depth)
for i, (code, val, ln) in enumerate(pairs):
    if code == 0 and val == 'SECTION' and i + 1 < len(pairs) and pairs[i + 1][0] == 2:
        sections.append(pairs[i + 1][1])
if not pairs or pairs[-1][1] != 'EOF':
    bad("the file does not end with EOF")
for need in ('HEADER', 'TABLES', 'BLOCKS', 'ENTITIES', 'OBJECTS'):
    if need not in sections:
        bad("no %s section: AutoCAD needs one in R2000" % need)

# ---------------------------------------------------------------- tables
# Every symbol table stops at AcDbSymbolTable -- except DIMSTYLE, which carries
# a second marker and its own count. That single exception is the bug this file
# was written for.
RECORD_CLASS = {
    'VPORT': 'AcDbViewportTableRecord',
    'LTYPE': 'AcDbLinetypeTableRecord',
    'LAYER': 'AcDbLayerTableRecord',
    'STYLE': 'AcDbTextStyleTableRecord',
    'VIEW': 'AcDbViewTableRecord',
    'UCS': 'AcDbUCSTableRecord',
    'APPID': 'AcDbRegAppTableRecord',
    'DIMSTYLE': 'AcDbDimStyleTableRecord',
    'BLOCK_RECORD': 'AcDbBlockTableRecord',
}
tables_seen = []
i = 0
while i < len(pairs):
    code, val, ln = pairs[i]
    if code == 0 and val == 'TABLE':
        name = pairs[i + 1][1] if i + 1 < len(pairs) and pairs[i + 1][0] == 2 else '?'
        tables_seen.append(name)
        head, j = [], i + 2
        while j < len(pairs) and pairs[j][0] != 0:
            head.append(pairs[j])
            j += 1
        flat = [(c, v) for c, v, _ in head]
        if (100, 'AcDbSymbolTable') not in flat:
            bad("table %s has no 100 AcDbSymbolTable in its header" % name, ln)
        if name == 'DIMSTYLE':
            if (100, 'AcDbDimStyleTable') not in flat:
                bad("the DIMSTYLE table header is missing 100 AcDbDimStyleTable -- "
                    "AutoCAD discards the drawing over this one", ln)
            elif not any(c == 71 for c, _ in flat):
                bad("the DIMSTYLE table header has the class marker but no 71 count", ln)
            else:
                k70 = [n for n, (c, _) in enumerate(flat) if c == 70]
                kcl = [n for n, (c, v) in enumerate(flat) if c == 100 and v == 'AcDbDimStyleTable']
                if k70 and kcl and kcl[0] < k70[0]:
                    bad("the DIMSTYLE class marker comes before the 70 count", ln)
        else:
            extra = [v for c, v in flat if c == 100 and v != 'AcDbSymbolTable']
            if extra:
                bad("table %s carries a marker only DIMSTYLE may have: %s"
                    % (name, ', '.join(extra)), ln)
        # the records inside it
        while j < len(pairs):
            c2, v2, l2 = pairs[j]
            if c2 == 0 and v2 == 'ENDTAB':
                break
            if c2 == 0 and v2 == name:
                rec, k = [], j + 1
                while k < len(pairs) and pairs[k][0] != 0:
                    rec.append(pairs[k])
                    k += 1
                fr = [(c, v) for c, v, _ in rec]
                want = RECORD_CLASS.get(name)
                if want and (100, want) not in fr:
                    bad("a %s record has no 100 %s" % (name, want), l2)
                if (100, 'AcDbSymbolTableRecord') not in fr:
                    bad("a %s record has no 100 AcDbSymbolTableRecord" % name, l2)
                # DIMSTYLE alone is handled on 105; everything else on 5
                hcode = 105 if name == 'DIMSTYLE' else 5
                if not any(c == hcode for c, _ in fr):
                    bad("a %s record has no handle on group code %d" % (name, hcode), l2)
                if name == 'DIMSTYLE' and any(c == 5 for c, _ in fr):
                    bad("a DIMSTYLE record is handled on 5; that table uses 105", l2)
            j += 1
        i = j
    i += 1
for need in RECORD_CLASS:
    if need not in tables_seen:
        bad("no %s table" % need)

# ---------------------------------------------------------------- handles
handles, dupes = set(), []
for n, (code, val, ln) in enumerate(pairs):
    if code not in (5, 105):
        continue
    prev = pairs[n - 1] if n else (None, None, None)
    if prev[0] == 0 and prev[1] == 'SECTION':
        continue
    # $HANDSEED carries its value on group code 5 as well, and it is a seed
    # rather than a handle -- counting it makes the file look like it uses a
    # handle it has not used, and then fails its own seed check.
    if prev[0] == 9:
        continue
    h = val.strip().upper()
    if h in ('', '0'):
        bad("a handle is %r" % val, ln)
        continue
    try:
        int(h, 16)
    except ValueError:
        bad("handle %r is not hexadecimal" % val, ln)
        continue
    if h in handles:
        dupes.append((h, ln))
    handles.add(h)
for h, ln in dupes[:6]:
    bad("handle %s is used twice" % h, ln)

seed = None
for n, (code, val, ln) in enumerate(pairs):
    if code == 9 and val == '$HANDSEED' and n + 1 < len(pairs):
        seed = pairs[n + 1][1].strip()
if seed is None:
    bad("$HANDSEED is not in the header")
elif handles:
    try:
        if int(seed, 16) <= max(int(h, 16) for h in handles):
            bad("$HANDSEED %s is not above every handle in the file" % seed)
    except ValueError:
        bad("$HANDSEED %r is not hexadecimal" % seed)

# ---------------------------------------------------------------- entities
# Every graphical object carries AcDbEntity and then its own subclass, and
# names a layer that exists.
layer_names = set()
i = 0
while i < len(pairs):
    code, val, ln = pairs[i]
    if code == 0 and val == 'TABLE' and pairs[i + 1][1] == 'LAYER':
        j = i + 2
        while j < len(pairs) and not (pairs[j][0] == 0 and pairs[j][1] == 'ENDTAB'):
            if pairs[j][0] == 0 and pairs[j][1] == 'LAYER':
                k = j + 1
                while k < len(pairs) and pairs[k][0] != 0:
                    if pairs[k][0] == 2:
                        layer_names.add(pairs[k][1])
                        break
                    k += 1
            j += 1
        break
    i += 1

GRAPHICAL = {'LINE', 'LWPOLYLINE', 'POLYLINE', 'CIRCLE', 'ARC', 'ELLIPSE', 'SPLINE',
             'POINT', 'TEXT', 'MTEXT', 'INSERT', 'HATCH', 'SOLID', 'DIMENSION',
             'LEADER', 'XLINE', 'RAY', 'ATTDEF', 'ATTRIB', 'VIEWPORT'}
missing_layer, no_entity_class, blocks_used, block_names = collections.Counter(), collections.Counter(), set(), set()
in_entities = False
for n, (code, val, ln) in enumerate(pairs):
    if code == 2 and n and pairs[n - 1][0] == 0 and pairs[n - 1][1] == 'SECTION':
        in_entities = val in ('ENTITIES', 'BLOCKS')
    if code == 0 and val == 'BLOCK':
        k = n + 1
        while k < len(pairs) and pairs[k][0] != 0:
            if pairs[k][0] == 2:
                block_names.add(pairs[k][1])
                break
            k += 1
    if code != 0 or val not in GRAPHICAL or not in_entities:
        continue
    body, k = [], n + 1
    while k < len(pairs) and pairs[k][0] != 0:
        body.append((pairs[k][0], pairs[k][1]))
        k += 1
    if (100, 'AcDbEntity') not in body:
        no_entity_class[val] += 1
    lay = [v for c, v in body if c == 8]
    if lay and layer_names and lay[0] not in layer_names:
        missing_layer[lay[0]] += 1
    if val == 'INSERT':
        nm = [v for c, v in body if c == 2]
        if nm:
            blocks_used.add(nm[0])
for t, n in no_entity_class.most_common(5):
    bad("%d %s entit%s without 100 AcDbEntity" % (n, t, 'y' if n == 1 else 'ies'))
for lay, n in missing_layer.most_common(5):
    bad("%d entities are on layer %r, which is not in the LAYER table" % (n, lay))
for b in sorted(blocks_used - block_names):
    bad("an INSERT names block %r, which is not defined" % b)

# ---------------------------------------------------------------- verdict
print("file    :", path)
print("lines   :", len(lines), " tags:", len(pairs))
print("sections:", ', '.join(sections))
print("tables  :", ', '.join(tables_seen))
print("handles :", len(handles), " seed:", seed)
if problems:
    print()
    print("PROBLEMS a strict reader would raise:")
    for p in problems:
        print("   -", p)
print()
print("RESULT:", "FAIL" if problems else "PASS")
sys.exit(1 if problems else 0)
